algorithms/src/std_algorithms/Kokkos_Constraints.hpp

```diff
@@ -0,0 +1,268 @@
+                                are_random_access_iterators<Tail...>::value;
+};
+
+//
+// iterators_are_accessible_from
+//
+template <class... Args>
+struct iterators_are_accessible_from;
+
+template <class ExeSpace, class IteratorType>
+struct iterators_are_accessible_from<ExeSpace, IteratorType> {
+  using view_type = typename IteratorType::view_type;
+  static constexpr bool value =
+      SpaceAccessibility<ExeSpace,
+                         typename view_type::memory_space>::accessible;
+};
+
+template <class ExeSpace, class Head, class... Tail>
+struct iterators_are_accessible_from<ExeSpace, Head, Tail...> {
+  static constexpr bool value =
+      iterators_are_accessible_from<ExeSpace, Head>::value &&
+      iterators_are_accessible_from<ExeSpace, Tail...>::value;
+};
+
+template <class ExecutionSpace, class IteratorType>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_random_access_and_accessible(const ExecutionSpace& /* ex */,
+                                           IteratorType /* it */) {
+  static_assert(
+      are_random_access_iterators<IteratorType>::value,
+      "Currently, Kokkos standard algorithms require random access iterators.");
+  static_assert(
+      iterators_are_accessible_from<ExecutionSpace, IteratorType>::value,
+      "Incompatible view/iterator and execution space");
+}
+
+template <class ExecutionSpace, class IteratorType1, class IteratorType2>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_random_access_and_accessible(const ExecutionSpace& ex,
+                                           IteratorType1 it1,
+                                           IteratorType2 it2) {
+  static_assert_random_access_and_accessible(ex, it1);
+  static_assert_random_access_and_accessible(ex, it2);
+}
+
+template <class ExecutionSpace, class IteratorType1, class IteratorType2,
+          class IteratorType3>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_random_access_and_accessible(const ExecutionSpace& ex,
+                                           IteratorType1 it1, IteratorType2 it2,
+                                           IteratorType3 it3) {
+  static_assert_random_access_and_accessible(ex, it1);
+  static_assert_random_access_and_accessible(ex, it2);
+  static_assert_random_access_and_accessible(ex, it3);
+}
+
+template <class ExecutionSpace, class IteratorType1, class IteratorType2,
+          class IteratorType3, class IteratorType4>
+KOKKOS_INLINE_FUNCTION constexpr void
+static_assert_random_access_and_accessible(const ExecutionSpace& ex,
+                                           IteratorType1 it1, IteratorType2 it2,
+                                           IteratorType3 it3,
+                                           IteratorType4 it4) {
+  static_assert_random_access_and_accessible(ex, it1);
+  static_assert_random_access_and_accessible(ex, it2);
+  static_assert_random_access_and_accessible(ex, it3);
+  static_assert_random_access_and_accessible(ex, it4);
+}
```

Comment:
Is there a good reason not to write this as
```suggestion
template <class ExecutionSpace, class... IteratorTypes>
KOKKOS_INLINE_FUNCTION constexpr void
static_assert_random_access_and_accessible(const ExecutionSpace& /* ex */,
                                           IteratorTypes...) {
  static_assert(
      are_random_access_iterators<IteratorTypes...>::value,
      "Currently, Kokkos standard algorithms require random access iterators.");
  static_assert(
      iterators_are_accessible_from<ExecutionSpace, IteratorTypes...>::value,
      "Incompatible view/iterator and execution space");
}
```
?

Reply:
thanks! yes totally right, don't know how I missed that :)